history properties should throw SecurityError when not in a fully active Document

sideshowbarker opened the pull request:
#### e4cbd15de56a52641bcdd542c0276fd48c0bcc9d
<pre>
history properties should throw SecurityError when not in a fully active Document
<a href="https://bugs.webkit.org/show_bug.cgi?id=264962">https://bugs.webkit.org/show_bug.cgi?id=264962</a>

Reviewed by Darin Adler and Ryosuke Niwa.

This change makes all History API methods and getters and setters throw
a SecurityError if called from a document that isn’t fully active (has
no Frame) — as required by the series of algorithms in the HTML spec
starting at <a href="https://html.spec.whatwg.org/multipage/#dom-history-length">https://html.spec.whatwg.org/multipage/#dom-history-length</a>

* LayoutTests/http/tests/dom/same-origin-detached-window-properties-expected.txt:
* LayoutTests/http/tests/dom/same-origin-detached-window-properties.html:
* LayoutTests/imported/w3c/web-platform-tests/html/browsers/history/the-history-interface/history_properties_only_fully_active-expected.txt:
* Source/WebCore/bindings/js/JSHistoryCustom.cpp:
(WebCore::JSHistory::state const):
* Source/WebCore/page/History.cpp:
(WebCore::isDocumentFullyActive):
(WebCore::documentNotFullyActive):
(WebCore::History::length const):
(WebCore::History::scrollRestoration const):
(WebCore::History::setScrollRestoration):
(WebCore::History::back):
(WebCore::History::forward):
(WebCore::History::go):
(WebCore::History::stateObjectAdded):
(WebCore::History::state): Deleted.
* Source/WebCore/page/History.h:

Canonical link: <a href="https://commits.webkit.org/274260@main">https://commits.webkit.org/274260@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/19896dbf4a63f5e12d8ac6705d88f38686903100

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/36752 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/15688 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/38990 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/39395 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/32925 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/37997 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/18173 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/12800 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/31495 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/37314 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/13220 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/32478 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/11564 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/11575 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/32755 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/40643 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/33290 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/33099 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/37494 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/11852 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/9669 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/35614 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/13516 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/32433 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/8642 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/12251 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/12748 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/WebCore/page/History.cpp:
```diff
@@ -58,11 +58,16 @@ History::History(LocalDOMWindow& window)
-unsigned History::length() const
+static Exception documentNotFullyActiveException()
+{
+    return Exception { ExceptionCode::SecurityError, "Attempt to use History API from a document that isn't fully active"_s };
```
Can you leave out `Exception` here?

sideshowbarker replied:
```
return Exception { ExceptionCode::SecurityError, "Attempt to use History API from a document that isn't fully active"_s };
```
> Can you leave out `Exception` here?

Is there a way to do that while also setting the "_Attempt to use History API from a document that isn't fully active"_ custom message?

Or do you mean that custom message seems unnecessary, and it’s sufficient to just throw a `SecurityError` without any custom message?

Ahmad-S792 replied:
It might be to rename function to `documentNotFullyActive` without word 'Exception' because it is explicit from function type.

sideshowbarker replied:
> It might be to rename function to `documentNotFullyActive` without word 'Exception' because it is explicit from function type.

aha — thanks, yes, changed it now to just `documentNotFullyActive`

darinadler replied:
I meant to ask a question about something else, much simpler. Can the return statement just be `return { … }` instead of `return Exception { … }`?

darinadler replied:
The rename also seems fine.

sideshowbarker replied:
> I meant to ask a question about something else, much simpler. Can the return statement just be `return { … }` instead of `return Exception { … }`?

ah OK, I see now. So when I try that, I get a _“chosen constructor is explicit in copy-initialization”_ error from clang:
```
/opt/WebKit/Source/WebCore/page/History.cpp:68:12: error: chosen constructor is explicit in copy-initialization
    return { ExceptionCode::SecurityError, "Attempt to use History API from a document that isn't fully active"_s };
           ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
...
/opt/WebKit/Source/WebCore/dom/Exception.h:50:19: note: explicit constructor declared here
inline Exception::Exception(ExceptionCode code, String message)
                  ^
```
And looking at the source in `dom/Exception.h`, that seems to make sense — though the actual precise cause is the `explicit` keyword up at line 36:
https://github.com/WebKit/WebKit/blob/8a9ddac3f43f8b4bd95d3427a86612e216947c63/Source/WebCore/dom/Exception.h#L34-L36

So it seems the compiler’s doing what it’s been told to do, and would only allow `return { ExceptionCode::SecurityError, … }` if that constructor weren’t intentionally declared `explicit`?

darinadler replied:
Yes that’s fine. So the answer is no, can’t leave it out.

darinadler commented on Source/WebCore/page/History.cpp:
```diff
@@ -58,11 +58,16 @@ History::History(LocalDOMWindow& window)
     auto* frame = this->frame();
-    if (!frame)
-        return 0;
+    if (!frame || !frame->document()->isFullyActive())
```
Seems like a helper function for this would make the code more compact and less repetitive. And a nice place to put a null check for the document if we later find out it’s needed.
```
static bool isDocumentFullyActive(Frame* frame)
{
    return frame && frame->document()->isFullActive();
}
```
```
    if (!isDocumentFullyActive(frame))
```

sideshowbarker replied:
Thanks — added a `isDocumentFullyActive(LocalFrame* frame)` helper function

darinadler commented on Source/WebCore/page/History.cpp:
```diff
@@ -129,51 +137,65 @@ bool History::isSameAsCurrentState(SerializedScriptValue* state) const
+ExceptionOr<void> History::back()
 {
-    go(-1);
+    auto voidOrException = go(-1);
```
Just `return go(-1)`

sideshowbarker replied:
Reduced all the `ExceptionOr` with `go(…)` cases to the appropriate `return go(…)` throughout

darinadler commented on Source/WebCore/page/History.cpp:
```diff
@@ -129,51 +137,65 @@ bool History::isSameAsCurrentState(SerializedScriptValue* state) const
+ExceptionOr<void> History::forward()
 {
-    go(1);
+    auto voidOrException = go(1);
```
Just `return go(1)`

darinadler commented on Source/WebCore/page/History.cpp:
```diff
@@ -129,51 +137,65 @@ bool History::isSameAsCurrentState(SerializedScriptValue* state) const
+ExceptionOr<void> History::forward(Document& document)
 {
-    go(document, 1);
+    auto voidOrException = go(document, 1);
```
Just `return go(document, 1)`

darinadler commented on Source/WebCore/page/History.cpp:
```diff
@@ -194,7 +216,9 @@ ExceptionOr<void> History::stateObjectAdded(RefPtr<SerializedScriptValue>&& data
 
     auto* frame = this->frame();
-    if (!frame || !frame->page())
+    if (!frame)
```
Why no `document()->isFullyActive()` check here?

sideshowbarker replied:
```
    if (!frame)
```
> Why no `document()->isFullyActive()` check here?

I seem to have missed that one case — thanks for catching it; fixed now

rniwa reviewed (approved):
r=me once comments made by Darin are addressed.

sideshowbarker committed:
history properties should throw SecurityError when not in a fully active Document
https://bugs.webkit.org/show_bug.cgi?id=264962

Reviewed by Darin Adler and Ryosuke Niwa.

This change makes all History API methods and getters and setters throw
a SecurityError if called from a document that isn’t fully active (has
no Frame) — as required by the series of algorithms in the HTML spec
starting at https://html.spec.whatwg.org/multipage/#dom-history-length

* LayoutTests/http/tests/dom/same-origin-detached-window-properties-expected.txt:
* LayoutTests/http/tests/dom/same-origin-detached-window-properties.html:
* LayoutTests/imported/w3c/web-platform-tests/html/browsers/history/the-history-interface/history_properties_only_fully_active-expected.txt:
* Source/WebCore/bindings/js/JSHistoryCustom.cpp:
(WebCore::JSHistory::state const):
* Source/WebCore/page/History.cpp:
(WebCore::isDocumentFullyActive):
(WebCore::documentNotFullyActive):
(WebCore::History::length const):
(WebCore::History::scrollRestoration const):
(WebCore::History::setScrollRestoration):
(WebCore::History::back):
(WebCore::History::forward):
(WebCore::History::go):
(WebCore::History::stateObjectAdded):
(WebCore::History::state): Deleted.
* Source/WebCore/page/History.h:

Canonical link: https://commits.webkit.org/274260@main
